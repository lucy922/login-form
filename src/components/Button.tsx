import { HTMLAttributes, ReactNode } from "react";
import classes from "../classes/Button.classes";
import classNames from "classnames";

export interface IButtonClasses {
  base: string;
  variant: {
    primary: string;
    secondary: string;
  };
  size: {
    sm: string;
    md: string;
  };
}

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  variant?: keyof IButtonClasses["variant"];
  size?: keyof IButtonClasses["size"];
  disabled?: boolean;
  name?: string;
  type?: "button" | "submit" | "reset";
  onClick?: () => void;
}

export const Button = (props: ButtonProps) => {
  return (
    <button
      {...props}
      type="submit"
      onClick={props.onClick}
      name="button"
      className={classNames(
        classes.base,
        classes.variant[props.variant === "primary" ? "primary" : "secondary"],
        classes.size[props.size === "sm" ? "sm" : "md"]
      )}
    >
      {props.children}
    </button>
  );
};
