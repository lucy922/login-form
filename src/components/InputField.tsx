import * as React from 'react'
import classes from '../classes/InputField.classes';

 export interface IinputFieldClasses{
   base: string
}

 export interface Props extends React.HTMLAttributes<HTMLInputElement>{
  type?: string,
  placeholder?: string,
  name?: string,
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void ,
  value?: string
}



const InputField = (props: Props) => {
  return (
       <input {...props}
            type={props.type}
            placeholder={props.placeholder}
            name={props.name}
            onChange={props.onChange}
            value={props.value}
            className={classes.base}
          />
  )
}

export default InputField