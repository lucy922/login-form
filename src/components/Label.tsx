
interface Props{
  children: string,
  className: string,
  color?: string
  htmlFor?: string}

const Label = (props: Props) => {
  return (
    <label htmlFor={props.htmlFor} className={props.className} color={props.color}>{props.children}</label>
  )
}

export default Label