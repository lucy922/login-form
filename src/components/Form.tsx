import React, { useState } from "react";
import { User } from "../user";
import { Button } from "./Button";
import InputField from "./InputField";
import Label from "./Label";

const Form = () => {
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [user, setUser] = useState<User[]>([]);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    setUser([{ email, password }]);
    setEmail("");
    setPassword("");
    console.log(user);
  };

  return (
    <form
    data-testid="login-form"
      onSubmit={(e) => handleSubmit(e)}
      className="block max-w-lg mx-auto bg-white drop-shadow-md rounded mt-12 p-4"
    >
      <h1 className="text-3xl pb-3">Login Form</h1>
      <div className="pt-5">
        <div className="pt-3">
          <Label className="block text-start pb-1">Email</Label>
          <InputField
          data-testid="email-input"
            type="text"
            placeholder="Enter email.."
            name="email"
            onChange={(e) => setEmail(e.target.value)}
            value={email}
          />
        </div>
        <div className=" pt-7">
          <Label className="block text-start pb-1">Password</Label>
          <InputField
          data-testid="password-input"
            type="password"
            placeholder="Enter password"
            name="password"
            onChange={(e) => setPassword(e.target.value)}
            value={password}
          />
        </div>
      </div>
      <Button type="submit" name="button" variant="primary" size="sm" className="mt-6">Login</Button>
    </form>
  );
};

export default Form;
