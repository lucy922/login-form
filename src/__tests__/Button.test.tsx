import { Button } from "../components/Button";
import { ButtonProps } from "../components/Button";
import { create, } from 'react-test-renderer'
import { fireEvent, render, screen} from "@testing-library/react";
import classNames from "classnames";
import classes from '../classes/Button.classes'


  describe('Button component', () => {
    it('should render button correctly', () => {
      render(<Button>Login</Button>);
      expect(screen.getByRole('button')).toBeInTheDocument()
    })

    it('should fire onClick event', () => {
      const onClick = jest.fn()
      render(<Button onClick={onClick} type="submit"/>)
      const button = screen.getByRole('button')
      fireEvent.click(button)
      expect(onClick).toHaveBeenCalled()
    })

    it('Button text', () => {
      render(<Button>Login</Button>);
      expect(screen.getByRole('button')).toHaveTextContent('Login')
    })

    it('should be enabled', () => {
      render(<Button>Login</Button>);
      expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
    })

    it('should be disabled', () => {
      render(<Button disabled={true}>Login</Button>);
      expect(screen.getByRole('button')).toBeDisabled()
    })

    it('Matches snapshot', () => {
      const tree = create(<Button>Login</Button>).toJSON();
      expect(tree).toMatchSnapshot();
    })

      it('should render primary button', () => {
        const variantProps: ButtonProps = {
          variant: 'primary'
        }
        render(<Button {...variantProps}>Login</Button>);
        expect(screen.getByRole('button')).toHaveClass(classNames(classes.variant.primary))
        const tree = create(<Button {...variantProps}>primary</Button>).toJSON();
        expect(tree).toMatchSnapshot();
      })
      
      it('should render secondary button', () => {
        const variantProps: ButtonProps = {
          variant: 'secondary'
        }
        render(<Button {...variantProps}>Login</Button>);
        expect(screen.getByRole('button')).toHaveClass(classNames(classes.variant.secondary))
        const tree = create(<Button {...variantProps}>secondary</Button>).toJSON();
        expect(tree).toMatchSnapshot();
      })
  
      it('should render size sm', () => {
        const sizeProps: ButtonProps = {
          size: 'sm'
        }
        render(<Button {...sizeProps}>Login</Button>);
        expect(screen.getByRole('button')).toHaveClass(classNames(classes.size.sm))
        const tree = create(<Button {...sizeProps}>small</Button>).toJSON();
        expect(tree).toMatchSnapshot()
      })
      
      it('should render size md', () => {
        const sizeProps: ButtonProps = {
          size: 'md'
        }
        render(<Button {...sizeProps}>Login</Button>);
        expect(screen.getByRole('button')).toHaveClass(classNames(classes.size.md))
        const tree = create(<Button {...sizeProps}>medium</Button>).toJSON();
        expect(tree).toMatchSnapshot()
      })
    })
    


  
 

 


