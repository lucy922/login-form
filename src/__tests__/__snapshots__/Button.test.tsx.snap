// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches snapshot 1`] = `
<button
  className="p-2 rounded text-center text-white bg-red-800  w-32"
  name="button"
  type="submit"
>
  Login
</button>
`;

exports[`Button component should render primary button 1`] = `
<button
  className="p-2 rounded text-center text-white bg-blue-800  w-32"
  name="button"
  type="submit"
  variant="primary"
>
  primary
</button>
`;

exports[`Button component should render secondary button 1`] = `
<button
  className="p-2 rounded text-center text-white bg-red-800  w-32"
  name="button"
  type="submit"
  variant="secondary"
>
  secondary
</button>
`;

exports[`Button component should render size md 1`] = `
<button
  className="p-2 rounded text-center text-white bg-red-800  w-32"
  name="button"
  size="md"
  type="submit"
>
  medium
</button>
`;

exports[`Button component should render size sm 1`] = `
<button
  className="p-2 rounded text-center text-white bg-red-800 w-24"
  name="button"
  size="sm"
  type="submit"
>
  small
</button>
`;
